### PR TITLE
Splunk token authentication

### PR DESCRIPTION
**Pull request by guyfreund**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/33995

## Description
Added support for Splunk token authentication. Based on https://github.com/splunk/splunk-sdk-python/pull/282

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [x] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Documentation

### CODE REVIEW

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_description.md [hunk 1]
```diff
@@ -1 +1,3 @@
 Use the SplunkPy integration to fetch incidents from Splunk ES, and query results by SID.
+
+To use Splunk token authentication, set the Username to `_token` and set the Password to `$TOKEN_VALUE`.
```
Comment (ShahafBenYakir) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_description.md:3]:
```suggestion
To use Splunk token authentication, enter the text: `_token` to the **Username** field and your token value in the **Password** field.
```
BTW why `_token` and not `token`?

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 2]
```diff
@@ -626,9 +626,16 @@ def splunk_edit_notable_event_command(proxy):
...
+    if username == '_token':
+        headers = {"Authorization": "Bearer {}".format(password)}
+        auth_req = requests.post(
+            baseurl + 'services/auth/login', data={'output_mode': 'json'}, headers=headers, verify=VERIFY_CERTIFICATE
+        )
+    else:
+        auth_req = requests.post(
+            baseurl + 'services/auth/login', data={'username': username, 'password': password, 'output_mode': 'json'},
+            verify=VERIFY_CERTIFICATE
+        )
```
Comment (ShahafBenYakir) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:638]:
```suggestion
    data={'output_mode': 'json'}
    headers = {}
    if username == '_token':
        headers = {"Authorization": "Bearer {}".format(password)}
    else:
        data={'username': username, 'password': password, 'output_mode': 'json'}
        
    auth_req = requests.post(
        baseurl + 'services/auth/login', data=data, headers=headers, verify=VERIFY_CERTIFICATE
    )
```

File: Packs/SplunkPy/Integrations/SplunkPy/README.md [hunk 3]
```diff
@@ -17,7 +17,7 @@ This integration was integrated and tested with Splunk v7.2.
 | **Parameter** | **Description** | **Required** |
 | --- | --- | --- |
 | host | The host name to the server, including the scheme (x.x.x.x). | True |
-| authentication | The username used for authentication. | True |
+| authentication | The username used for authentication. To use Splunk token authentication, set the Username to `_token` and set the Password to `$TOKEN_VALUE`. | True |
```
Comment (ShahafBenYakir) [inline Packs/SplunkPy/Integrations/SplunkPy/README.md:20]: need to fix here too

**Review by ShahafBenYakir — COMMENTED** [review]
Have we tested this?

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_description.md (hunk 1 above)
Comment (guyfreund) [reply Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_description.md:3]: as @glicht requested https://github.com/demisto/etc/issues/33995#issuecomment-800964142

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 4]
```diff
@@ -617,37 +621,30 @@ def splunk_submit_event_hec_command():
...
+                                        owner=demisto.get(demisto.args(), 'owner'), eventIDs=eventIDs,
+                                        auth_token=auth_token, sessionKey=sessionKey)
     if 'success' not in response_info or not response_info['success']:
         demisto.results({'ContentsFormat': formats['text'], 'Type': entryTypes['error'],
                          'Contents': "Could not update notable "
                                      "events: " + demisto.args()['eventIDs'] + ' : ' + str(response_info)})
-
-    demisto.results('Splunk ES Notable events: ' + response_info.get('message'))
+    else:
+        demisto.results('Splunk ES Notable events: ' + response_info.get('message'))
```
Comment (ShahafBenYakir) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:647]: Why have we chaged this?

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 5]
```diff
@@ -617,37 +621,30 @@ def splunk_submit_event_hec_command():
...
-    password = demisto.params()['authentication']['password']
-    auth_req = requests.post(baseurl + 'services/auth/login',
-                             data={'username': username, 'password': password, 'output_mode': 'json'},
-                             verify=VERIFY_CERTIFICATE)
+def splunk_edit_notable_event_command(proxy, service, auth_token):
+    params = demisto.params()
+    base_url = 'https://' + params['host'] + ':' + params['port'] + '/'
+    sessionKey = service.token if not auth_token else None
 
-    sessionKey = auth_req.json()['sessionKey']
```
Comment (ShahafBenYakir) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:633]: You are saying this is redundent?

Comment (guyfreund) [reply Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:633]: yes, this function was copied from https://gist.github.com/LukeMurphey/b45a5425572deffd2e955e460f261dbf but its a wrong use, we already authenticate in the `main` function, so no need to re-authenticate. The Splunk `Service` classes does this authentication in the `login` method:
image

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_description.md [hunk 6]
```diff
@@ -1 +1,3 @@
 Use the SplunkPy integration to fetch incidents from Splunk ES, and query results by SID.
+
+To use Splunk token authentication, enter the text: `_token` to the **Username** field and your token value to the **Password** field.
```
Comment (roysagi) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_description.md:3]: maybe you should add an explanation about how to get the token from splunk's side?

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py (hunk 4 above)
Comment (guyfreund) [reply Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:647]: changed this to a better implementation

**Comment by glicht** [comment]
Please see the PR form @amshamah419 : https://github.com/demisto/content/pull/11651 . I think there may be some overlapping here.

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 7]
```diff
@@ -115,25 +115,26 @@ def convert_to_str(obj):
...
     searchID -- An ID of a search. All of the events associated with this search will be modified
      unless a list of eventIDs are provided that limit the scope to a sub-set of the results.
+     auth_token - The authentication token to use
+     sessionKey -- The session key to use
     """
 
     # Make sure that the session ID was provided
-    if sessionKey is None:
-        raise Exception("A session key was not provided")
+    if sessionKey is None and not auth_token or auth_token is None and not sessionKey:
```
Comment (glicht) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:136]: why not do:
```suggestion
    if not sessionKey and not auth_token:
```

**Review by ShahafBenYakir — APPROVED** [review]

**Pull request merged**